Improve type hints in manager classes

mvdbeek opened the pull request:
Most of the return values of the central managers now have type annotations, which has already caught a couple of minor issues (https://github.com/galaxyproject/galaxy/commit/4fdcb1baefed02752369c7e86d244a2986b8c5c4, https://github.com/galaxyproject/galaxy/commit/77af0f6f95d661cd0e6985676b419937e778f7fc and https://github.com/galaxyproject/galaxy/commit/c8d4f1435ff95c97d740d24a80d692afea19f1c1)

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek committed:
Make most base managers generic and parameterize concrete item managers

This means that when we list/filter/retrieve items from a given manager
we get the correct return type, instead of `any`.

mvdbeek committed:
Split UserManager into shared base manager and galaxy-specific user manager

mvdbeek committed:
Add ToolShed specific user manager

mvdbeek committed:
Fix bugs and type hints reveealed by improved type hints

mvdbeek committed:
Ensure dataset on disk befoe access

mvdbeek committed:
Drop unnecessary type ignores

mvdbeek committed:
Make ApiKeyManager generic for galaxy or ts models

mvdbeek committed:
Fix and adjust model type optionality

The model optionality on the sqlalchemy side seems weird.

mvdbeek committed:
Cannot use union in lagom

mvdbeek committed:
Fix type annotation for api_keys.key

mvdbeek committed:
Fix nullability of TS api key

mvdbeek committed:
Fixup setup_users in celery rate limit integration test

mvdbeek committed:
Fix unit tests

davelopez commented on lib/galaxy/managers/datasets.py:
```diff
@@ -376,7 +381,7 @@ def delete(self, item, flush: bool = True, stop_job: bool = False, **kwargs):
         return item
 
-    def purge(self, dataset_assoc, flush=True):
+    def purge(self, item: DI, flush=True, **kwargs) -> DI:
```
What is the reason to have `kwargs` here and in many of these functions? :thinking:

mvdbeek replied:
To maintain the signature with the base class, mypy complains about the signature mismatch otherwise

davelopez replied:
It would be nice to check what is used as parameters, remove those `kwargs`, and add explicit ones at some point, but it is something that can be done later. This is a huge improvement already.

davelopez commented on lib/galaxy/managers/users.py:
```diff
@@ -377,9 +295,8 @@ def error_if_anonymous(self, user, msg="Log-in required", **kwargs):
         return user
 
-    def get_user_by_identity(self, identity):
+    def get_user_by_identity(self, identity) -> Optional[U]:
```
```suggestion
    def get_user_by_identity(self, identity: str) -> Optional[U]:
```

davelopez commented on lib/galaxy/managers/users.py:
```diff
@@ -635,22 +492,172 @@ def send_subscription_email(self, email):
             session.commit()
 
-    def get_or_create_remote_user(self, remote_user_email):
+    def _get_user_by_email_case_insensitive(self, session, email):
```
```suggestion
    def _get_user_by_email_case_insensitive(self, session, email: str):
```

davelopez commented on lib/galaxy/model/__init__.py:
```diff
@@ -732,7 +732,7 @@ class User(Base, Dictifiable, RepresentById):
     update_time: Mapped[datetime] = mapped_column(default=now, onupdate=now, nullable=True)
     email: Mapped[str] = mapped_column(TrimmedString(255), index=True)
-    username: Mapped[Optional[str]] = mapped_column(TrimmedString(255), index=True, unique=True)
+    username: Mapped[str] = mapped_column(TrimmedString(255), index=True, unique=True)
```
If I understood correctly the third bullet point in https://github.com/galaxyproject/galaxy/pull/17958, we need to add `nullable=True` to `mapped_column` to reflect the real table definition.

The same applies to the rest of the model attributes that changed to non-optional type hints here.

mvdbeek replied:
Hehe, I was looking for that, it would have saved me from changing all those unit tests 😆

davelopez reviewed (commented):
Wonderful!
Feel free to ignore some of the type hint suggestions if they create more issues than they are worth.

mvdbeek committed:
Fix up history pruner unit test

mvdbeek committed:
Tighten up use of DatasetProtocol and DatasetHasHidProtocol

mvdbeek committed:
Make AccessibleManagerMixin generic

mvdbeek committed:
Specialize generated query in ModelManager

mvdbeek committed:
Add type parameters to DatasetManager

mvdbeek committed:
Add type annotation to HDAManager copy signature

mvdbeek committed:
Eliminate unnecessary type ignores

mvdbeek committed:
Undo DatasetSourceType change

mvdbeek committed:
Add type hint for dataset_manager_by_type

mvdbeek committed:
Turn DatasetSourceType into literal

This allows type lookups using TypedDict, which isn't possible with enum
members.

mvdbeek committed:
Do proper type narrowing in delete_batch

mvdbeek committed:
Ensure dataset should be on disk before opening dataset

mvdbeek committed:
Remove unnecessary by_id implementation

mvdbeek committed:
Add API test to ensure that a proper error is returned if the dataset can't be opened

mvdbeek committed:
Add more type hints in manager signatures

Co-authored-by: David López <46503462+davelopez@users.noreply.github.com>

mvdbeek committed:
Restore nullability of column definition

davelopez commented on lib/galaxy/model/__init__.py:
```diff
@@ -6834,10 +6844,10 @@ class HistoryDatasetCollectionAssociation(
+    history_id: Mapped[int] = mapped_column(ForeignKey("history.id"), index=True)
     name: Mapped[Optional[str]] = mapped_column(TrimmedString(255))
-    hid: Mapped[Optional[int]]
+    hid: Mapped[int] = mapped_column(nullabled=True)
```
```suggestion
    hid: Mapped[int] = mapped_column(nullable=True)
```

mvdbeek replied:
Ooops, thank you!

mvdbeek committed:
Fix nullable typo

Co-authored-by: David López <46503462+davelopez@users.noreply.github.com>

davelopez commented on lib/galaxy/managers/users.py:
```diff
@@ -635,22 +489,172 @@ def send_subscription_email(self, email):
+            raise exceptions.MessageException(f"User '{item.email}' has not been deleted, so they cannot be purged.")
+        private_role = self.app.security_agent.get_private_user_role(item)
+        # Delete History
+        for active_history in item.active_histories:
```
I've just noticed that this probably will not purge archived histories from the user... :grimacing: :disappointed:

davelopez commented on lib/galaxy/schema/schema.py:
```diff
@@ -699,11 +697,14 @@ class HDAInaccessible(HDACommon):
-    description="Whether this dataset belongs to a history (HDA) or a library (LDDA).",
-)
+HdaLddaField = Annotated[
+    DataItemSourceType,
```
I think what we want here is `DatasetSourceType` :thinking: 
```suggestion
    DatasetSourceType,
```

davelopez commented on lib/galaxy/webapps/galaxy/services/datasets.py:
```diff
@@ -606,7 +610,15 @@ def display(
         try:
-            dataset_instance = self.dataset_manager_by_type[hda_ldda].get_accessible(dataset_id, trans.user)
+            dataset_instance: Union[model.HistoryDatasetAssociation, model.LibraryDatasetDatasetAssociation]
+            if hda_ldda == "hda":
+                hda_manager = self.dataset_manager_by_type[hda_ldda]
+                dataset_instance = hda = hda_manager.get_accessible(dataset_id, trans.user)
+                hda_manager.ensure_dataset_on_disk(trans, hda)
+            else:
+                ldda_manager = self.dataset_manager_by_type[hda_ldda]
+                dataset_instance = ldda = ldda_manager.get_accessible(dataset_id, trans.user)
+                ldda_manager.ensure_dataset_on_disk(trans, ldda)
```
Why not something like this?
```suggestion
            dataset_instance: Union[model.HistoryDatasetAssociation, model.LibraryDatasetDatasetAssociation]
            manager = self.dataset_manager_by_type[hda_ldda]
            dataset_instance = manager.get_accessible(dataset_id, trans.user)
            manager.ensure_dataset_on_disk(trans, dataset_instance)
```

mvdbeek replied:
I think it's a covariance issue. I agree this is much cleaner, but mypy will complain that either the HDA or the LDDA manager can't work with the oppposite half of the HDA/LDDA union.

davelopez replied:
Ahh I see now... what a pitty 😕
I guess all methods should work with `DatasetInstance` but then I'm sure we will lose concrete typing in other places.

davelopez commented on lib/galaxy/webapps/galaxy/services/datasets.py:
```diff
@@ -720,15 +732,23 @@ def delete_batch(
+                if dataset.src == "hda":
+                    hda_manager = self.dataset_manager_by_type[dataset.src]
+                    hda = hda_manager.get_owned(dataset.id, trans.user)
+                    hda_manager.error_if_uploading(hda)
+                    hda_manager.error_unless_mutable(hda.history)
```
I wonder if one more dictionary called `raise_delete_error_by_type` that will call these 2 methods for `hda`s and just no-op for `ldda`s will eliminate this duplication :thinking:

davelopez reviewed (commented):
I love all these refactorings! thank you very much!

Just a couple of comments, one I think might be an ugly bug about archived (non-purged) histories :disappointed:
